sdk/lib/rtl/rangelist.c

```diff
@@ -47,237 +509,675 @@ typedef struct _RTL_RANGE_ENTRY
+
+        Iterator->MergedHead = NULL;
+    }
+
+    if (&NextRtlEntry->ListEntry == Iterator->RangeListHead)
+    {
+        DPRINT("RtlGetNextRange: return STATUS_NO_MORE_ENTRIES\n");
+        Iterator->Current = NULL;
+        *OutRange = NULL;
+        return STATUS_NO_MORE_ENTRIES;
+    }
+
+    if (NextRtlEntry->PrivateFlags & RTLP_ENTRY_IS_MERGED)
+    {
+        ASSERT(!Iterator->MergedHead);
+        Iterator->MergedHead = &NextRtlEntry->Merged.ListHead;
+
+        if (MoveForwards)
+        {
+            ListEntry = NextRtlEntry->Merged.ListHead.Flink;
+        }
+        else
+        {
+            ListEntry = NextRtlEntry->Merged.ListHead.Blink;
+        }
+    }
+    else
+    {
+        ListEntry = &NextRtlEntry->ListEntry;
+    }
+
+    NextRtlEntry = RtlpEntryFromLink(ListEntry);
+    DPRINT("RtlGetNextRange: %p, %p, *OutRange %X [%I64X-%I64X]\n", Iterator->Current, Iterator->MergedHead, NextRtlEntry, NextRtlEntry->Start, NextRtlEntry->End);
+
+    Iterator->Current = NextRtlEntry;
+    *OutRange = (PRTL_RANGE)NextRtlEntry;
+
+    return STATUS_SUCCESS;
+}
+
+BOOLEAN
+NTAPI
+RtlpIsRangeAvailable(
+    _Inout_ PRTL_RANGE_LIST_ITERATOR Iterator,
+    _In_ ULONGLONG Start,
+    _In_ ULONGLONG End,
+    _In_ UCHAR AttributeAvailableMask,
+    _In_ BOOLEAN Flag1,
+    _In_ BOOLEAN Flag2,
+    _In_ BOOLEAN MoveForwards,
+    _In_ PVOID Context,
+    _In_ PRTL_CONFLICT_RANGE_CALLBACK Callback)
+{
+    PRTL_RANGE RtlRange;
+
+    PAGED_CODE_RTL();
+    DPRINT("RtlpIsRangeAvailable: %p [%I64X-%I64X], %X, %X, Forwards %X, %p\n", Iterator, Start, End, Flag1, Flag2, MoveForwards, Callback);
+
+    ASSERT(Iterator);
+
+    for (RtlRange = Iterator->Current;
+         RtlRange;
+         RtlGetNextRange(Iterator, &RtlRange, MoveForwards))
+    {
+        if (MoveForwards)
+        {
+            if (!Iterator->MergedHead && End < RtlRange->Start)
+            {
+                return TRUE;
+            }
+        }
+        else if (!Iterator->MergedHead && Start > RtlRange->End)
+        {
+            return TRUE;
+        }
+
+        if ((RtlRange->Start > Start && RtlRange->Start > End) ||
+            (RtlRange->Start < Start && RtlRange->End < Start))
+        {
+            continue;
+        }
 
-    CopyRangeList->Flags = RangeList->Flags;
+        DPRINT("RtlpIsRangeAvailable: Intersection [%I64X-%I64X] and [%I64X-%I64X]\n", Start, End, RtlRange->Start, RtlRange->End);
 
-    Entry = RangeList->ListHead.Flink;
-    while (Entry != &RangeList->ListHead)
-    {
-        Current = CONTAINING_RECORD(Entry, RTL_RANGE_ENTRY, Entry);
+        if (Flag1 && (RtlRange->Flags & 1)) // FIXME
+        {
+            continue;
+        }
 
-        NewEntry = RtlpAllocateMemory(sizeof(RTL_RANGE_ENTRY), 'elRR');
-        if (NewEntry == NULL)
-            return STATUS_INSUFFICIENT_RESOURCES;
+        if ((AttributeAvailableMask & RtlRange->Attributes))
+        {
+            continue;
+        }
+
+        if (Flag2 == 0)
```

Comment:
Any idea what does this flag mean? I don't like putting an unknown logic like that, better not use it until we know what does it do

Reply:
Compatibility will be broken without these flags.